### ISSUE TITLE
NAN output after several epochs

### ISSUE THREAD

**Issue by fighting-liu** [mapillary/inplace_abn#4]
Thanks for this great work! I've used resnext101 pretrained model in my project, taking avg pooling output as embeddings and using triplet loss to train the model, everything works fine for hundreds of epochs, and the convergence works fine too. But the outputs become nan suddenly and the program crashed, any suggestions? Thanks!

**Comment by amrit110** [comment]
I get the same issue using the InPlaceABN in another custom model after several epochs. due to the nan output, the loss becomes nan and error manifests as illegal memory access. really hard to trace the issue, any ideas?

**Comment by pkontschieder** [comment]
Can you trace back this bug to our code, i.e. does it _not_ appear when you use standard pytorch BN + activation (leaky relu) layers?
In order to help you, we will need more information like a minimal example triggering this behavior, your GPU setting, console dumps, etc.

**Comment by pkontschieder** [comment]
@fighting-liu and @amrit110 
Do you have some follow-up information on this you can share with us?

**Comment by fighting-liu** [comment]
Sorry for reply late, as it's hard to reproduce the error, i need to run numbers of epochs to reproduce.  As the first time nan output occurs, i print the gradient of all parameters, all upstream gradients are ok until one of "InPlaceABN" layer, parameters are of shape (512,) (512,), and nan occurs in the gradient vector.
my setting: {"arch": "resnext101","activation": "leaky_relu","leaky_relu_slope": 0.01,"input_3x3": True,"bn_mode": "inplace","classes": 1000,"dilation": 1}
optimizer: Adam

**Comment by rotabulo** [comment]
@fighting-liu Could you please re-run the code by checking with an `assert` after [this line](https://github.com/mapillary/inplace_abn/blob/master/modules/functions.py#L108) whether `bias` contains 0s or nans ?

**Comment by fighting-liu** [comment]
@rotabulo  @pkontschieder  Yes, i have verified, `weight` tensor contains 0s after number of epochs. This is my code snippet
` if weight.eq(0).any(): raise`

**Comment by rotabulo** [comment]
@fighting-liu Thank you. We wrote about this potential issue in the paper, section 3.3 "Inversion of `\pi_{\gamma,\beta}`". In all trainings we have done, it never happened that gamma became 0. I am wondering whether you are using a l1-regularization of the network parameters and whether you are initializing all gamma to 1.

Nonetheless, a simple and easy fix that you can implement to move on with your training, consists in applying max(gamma,epsilon) to all scale parameters of inplace-abn right after the optimization step. This is equivalent to considering a constrained optimization where gamma should not become smaller than epsilon. We will then implement a proper fix. 

Let us know how it develops. Thanks.

**Comment by ducksoup** [comment]
@fighting-liu We implemented a possible fix in 1ffbf200e7dce680f15f046c6ba7c61386d43583 and 25ca7ce265bfe676280c461735b84e791b971310. Can you please check if this fixes your issue?

**Comment by fighting-liu** [comment]
@ducksoup Thanks, i will try it today, and further results will be reported here.

**Comment by fighting-liu** [comment]
@ducksoup @rotabulo Thanks, it works now!

**Comment by ducksoup** [comment]
@fighting-liu Thank you for reporting the bug, we are closing this now.
